substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/PosixJavaNetSubstitutions.java

```diff
@@ -119,84 +419,1339 @@ protected int peekData(DatagramPacket p) throws IOException {
+                                    }
+                                    //   860  #else
+                                } else {
+                                    //   862                           JNU_ThrowByName(env, JNU_JAVANETPKG "SocketException", "Socket closed");
+                                    throw new SocketException("Socket closed");
+                                    //   863  #endif
+                                }
+
+                                //   865                  } else if (ret == JVM_IO_INTR) {
+                            } else if (ret == Target_jvm.JVM_IO_INTR()) {
+                                //   866                      JNU_ThrowByName(env, JNU_JAVAIOPKG "InterruptedIOException",
+                                //   867                                      "operation interrupted");
+                                throw new InterruptedIOException("operation interrupted");
+                            }
+                        } finally {
+                            //   870                  if (mallocedPacket) {
+                            if (mallocedPacket) {
+                                //   871                      free(fullPacket);
+                                LibC.free(fullPacket);
+                                /* so that the outter try/finally does not attempt to double-free() */
+                                mallocedPacket = false;
+                            }
+                            //   874                  return;
+                            return;
+                        }
+                    }
+                }
+
+                //   878          len = SOCKADDR_LEN;
+                len_Pointer.write(JavaNetNetUtilMD.SOCKADDR_LEN());
+
+                //   879          n = NET_RecvFrom(fd, fullPacket, packetBufferLen, 0,
+                //   880                           (struct sockaddr *)&remote_addr, &len);
+
+                n = JavaNetNetUtilMD.NET_RecvFrom(fd, fullPacket, packetBufferLen, 0, remote_addr, len_Pointer);
+                //   881          /* truncate the data if the packet's length is too small */
+                //   882          if (n > packetBufferLen) {
+                if (n > packetBufferLen) {
+                    //   883              n = packetBufferLen;
+                    n = packetBufferLen;
+                }
+
+                //   885          if (n == JVM_IO_ERR) {
+                if (n == Target_jvm.JVM_IO_ERR()) {
+                    //   886              (*env)->SetIntField(env, packet, dp_offsetID, 0);
+                    Util_java_net_DatagramPacket.as_Target_java_net_DatagramPacket(packet).offset = 0;
+                    //   887              (*env)->SetIntField(env, packet, dp_lengthID, 0);
+                    Util_java_net_DatagramPacket.as_Target_java_net_DatagramPacket(packet).length = 0;
+                    if (Errno.errno() == Errno.ECONNREFUSED()) {
```

Comment:
How much better would we all sleep at night if you saved a copy of `Errno.errno()` after the call to `JavaNetNetUtilMD.NET_Recv` and used that for these tests, rather than assuming that none of the intervening calls did anything that might change that variable?  I don't *think* anything does, here, but it is easy enough to be safe.